upstream-sha256: 2d333cab2930d24e32eef76f91c7e27214f13cbc7d6cc3ea03de418911d9b04c
actual-sha256: 7aac49b5f2f7cdb000d6f134c159fa457451e41e04a46e0bbd92ed710305dd99
--- upstream.expected
+++ tolk.actual
@@ -7,125 +7,149 @@
    5 Ops.CAST            :            dtypes.int                               ['3']                            dtypes.int
    6 Ops.MUL             :            dtypes.int                               [3, 5]                           None
    7 Ops.INDEX           :            dtypes.float                             [0, 6]                           None
-   8 Ops.PARAM           :            dtypes.bool                              []                               ParamArg(1, dtypes.bool, 12, device='CPU')
-   9 Ops.CONST           :            dtypes.weakint                           []                               0
-  10 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
-  11 Ops.INDEX           :            dtypes.bool                              [8, 10]                          None
-  12 Ops.LOAD            :            dtypes.bool                              [11]                             None
-  13 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 24, device='CPU')
-  14 Ops.CONST           :            dtypes.weakint                           []                               12
-  15 Ops.CAST            :            dtypes.int                               ['12']                           dtypes.int
-  16 Ops.MUL             :            dtypes.int                               [3, 15]                          None
-  17 Ops.CONST           :            dtypes.weakint                           []                               4
-  18 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-  19 Ops.SHRINK          :            dtypes.float                             [13, 16, 18]                     None
-  20 Ops.LOAD            :            dtypes.float                             [19]                             None
-  21 Ops.INDEX           :            dtypes.float                             [20, 10]                         None
-  22 Ops.CONST           :            dtypes.weakfloat                         []                               -inf
-  23 Ops.CAST            :            dtypes.float                             ['-inf']                         dtypes.float
-  24 Ops.WHERE           :            dtypes.float                             [12, 21, 23]                     None
-  25 Ops.CONST           :            dtypes.weakint                           []                               1
-  26 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
-  27 Ops.INDEX           :            dtypes.bool                              [8, 26]                          None
-  28 Ops.LOAD            :            dtypes.bool                              [27]                             None
-  29 Ops.INDEX           :            dtypes.float                             [20, 26]                         None
-  30 Ops.WHERE           :            dtypes.float                             [28, 29, 23]                     None
-  31 Ops.CMPLT           :            dtypes.bool                              [24, 30]                         None
-  32 Ops.WHERE           :            dtypes.float                             [31, 30, 24]                     None
-  33 Ops.INDEX           :            dtypes.bool                              [8, 2]                           None
-  34 Ops.LOAD            :            dtypes.bool                              [33]                             None
-  35 Ops.INDEX           :            dtypes.float                             [20, 2]                          None
-  36 Ops.WHERE           :            dtypes.float                             [34, 35, 23]                     None
-  37 Ops.CMPLT           :            dtypes.bool                              [32, 36]                         None
-  38 Ops.WHERE           :            dtypes.float                             [37, 36, 32]                     None
-  39 Ops.INDEX           :            dtypes.bool                              [8, 5]                           None
-  40 Ops.LOAD            :            dtypes.bool                              [39]                             None
-  41 Ops.INDEX           :            dtypes.float                             [20, 5]                          None
-  42 Ops.WHERE           :            dtypes.float                             [40, 41, 23]                     None
-  43 Ops.CMPLT           :            dtypes.bool                              [38, 42]                         None
-  44 Ops.WHERE           :            dtypes.float                             [43, 42, 38]                     None
-  45 Ops.PARAM           :            dtypes.float                             []                               ParamArg(3, dtypes.float, 2, device='CPU')
-  46 Ops.INDEX           :            dtypes.float                             [45, 3]                          None
-  47 Ops.LOAD            :            dtypes.float                             [46]                             None
-  48 Ops.CMPLT           :            dtypes.bool                              [44, 47]                         None
-  49 Ops.WHERE           :            dtypes.float                             [48, 47, 44]                     None
-  50 Ops.STORE           :            dtypes.void                              [7, 49]                          None
-  51 Ops.ADD             :            dtypes.int                               [6, 26]                          None
-  52 Ops.INDEX           :            dtypes.float                             [0, 51]                          None
-  53 Ops.INDEX           :            dtypes.bool                              [8, 18]                          None
-  54 Ops.LOAD            :            dtypes.bool                              [53]                             None
-  55 Ops.ADD             :            dtypes.int                               [16, 18]                         None
-  56 Ops.SHRINK          :            dtypes.float                             [13, 55, 18]                     None
-  57 Ops.LOAD            :            dtypes.float                             [56]                             None
-  58 Ops.INDEX           :            dtypes.float                             [57, 10]                         None
-  59 Ops.WHERE           :            dtypes.float                             [54, 58, 23]                     None
-  60 Ops.CONST           :            dtypes.weakint                           []                               5
-  61 Ops.CAST            :            dtypes.int                               ['5']                            dtypes.int
-  62 Ops.INDEX           :            dtypes.bool                              [8, 61]                          None
-  63 Ops.LOAD            :            dtypes.bool                              [62]                             None
-  64 Ops.INDEX           :            dtypes.float                             [57, 26]                         None
-  65 Ops.WHERE           :            dtypes.float                             [63, 64, 23]                     None
-  66 Ops.CMPLT           :            dtypes.bool                              [59, 65]                         None
-  67 Ops.WHERE           :            dtypes.float                             [66, 65, 59]                     None
-  68 Ops.CONST           :            dtypes.weakint                           []                               6
-  69 Ops.CAST            :            dtypes.int                               ['6']                            dtypes.int
-  70 Ops.INDEX           :            dtypes.bool                              [8, 69]                          None
-  71 Ops.LOAD            :            dtypes.bool                              [70]                             None
-  72 Ops.INDEX           :            dtypes.float                             [57, 2]                          None
-  73 Ops.WHERE           :            dtypes.float                             [71, 72, 23]                     None
-  74 Ops.CMPLT           :            dtypes.bool                              [67, 73]                         None
-  75 Ops.WHERE           :            dtypes.float                             [74, 73, 67]                     None
-  76 Ops.CONST           :            dtypes.weakint                           []                               7
-  77 Ops.CAST            :            dtypes.int                               ['7']                            dtypes.int
-  78 Ops.INDEX           :            dtypes.bool                              [8, 77]                          None
+   8 Ops.PARAM           :            dtypes.float                             []                               ParamArg(3, dtypes.float, 2, device='CPU')
+   9 Ops.INDEX           :            dtypes.float                             [8, 3]                           None
+  10 Ops.LOAD            :            dtypes.float                             [9]                              None
+  11 Ops.PARAM           :            dtypes.bool                              []                               ParamArg(1, dtypes.bool, 12, device='CPU')
+  12 Ops.INDEX           :            dtypes.bool                              [11, 5]                          None
+  13 Ops.LOAD            :            dtypes.bool                              [12]                             None
+  14 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 24, device='CPU')
+  15 Ops.CONST           :            dtypes.weakint                           []                               12
+  16 Ops.CAST            :            dtypes.int                               ['12']                           dtypes.int
+  17 Ops.MUL             :            dtypes.int                               [3, 16]                          None
+  18 Ops.CONST           :            dtypes.weakint                           []                               4
+  19 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+  20 Ops.SHRINK          :            dtypes.float                             [14, 17, 19]                     None
+  21 Ops.LOAD            :            dtypes.float                             [20]                             None
+  22 Ops.INDEX           :            dtypes.float                             [21, 5]                          None
+  23 Ops.CONST           :            dtypes.weakfloat                         []                               -inf
+  24 Ops.CAST            :            dtypes.float                             ['-inf']                         dtypes.float
+  25 Ops.WHERE           :            dtypes.float                             [13, 22, 24]                     None
+  26 Ops.INDEX           :            dtypes.bool                              [11, 2]                          None
+  27 Ops.LOAD            :            dtypes.bool                              [26]                             None
+  28 Ops.INDEX           :            dtypes.float                             [21, 2]                          None
+  29 Ops.WHERE           :            dtypes.float                             [27, 28, 24]                     None
+  30 Ops.CONST           :            dtypes.weakint                           []                               1
+  31 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
+  32 Ops.INDEX           :            dtypes.bool                              [11, 31]                         None
+  33 Ops.LOAD            :            dtypes.bool                              [32]                             None
+  34 Ops.INDEX           :            dtypes.float                             [21, 31]                         None
+  35 Ops.WHERE           :            dtypes.float                             [33, 34, 24]                     None
+  36 Ops.CONST           :            dtypes.weakint                           []                               0
+  37 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
+  38 Ops.INDEX           :            dtypes.bool                              [11, 37]                         None
+  39 Ops.LOAD            :            dtypes.bool                              [38]                             None
+  40 Ops.INDEX           :            dtypes.float                             [21, 37]                         None
+  41 Ops.WHERE           :            dtypes.float                             [39, 40, 24]                     None
+  42 Ops.CMPLT           :            dtypes.bool                              [35, 41]                         None
+  43 Ops.CMPNE           :            dtypes.bool                              [41, 41]                         None
+  44 Ops.WHERE           :            dtypes.float                             [43, 41, 35]                     None
+  45 Ops.WHERE           :            dtypes.float                             [42, 41, 44]                     None
+  46 Ops.CMPLT           :            dtypes.bool                              [29, 45]                         None
+  47 Ops.CMPNE           :            dtypes.bool                              [45, 45]                         None
+  48 Ops.WHERE           :            dtypes.float                             [47, 45, 29]                     None
+  49 Ops.WHERE           :            dtypes.float                             [46, 45, 48]                     None
+  50 Ops.CMPLT           :            dtypes.bool                              [25, 49]                         None
+  51 Ops.CMPNE           :            dtypes.bool                              [49, 49]                         None
+  52 Ops.WHERE           :            dtypes.float                             [51, 49, 25]                     None
+  53 Ops.WHERE           :            dtypes.float                             [50, 49, 52]                     None
+  54 Ops.CMPLT           :            dtypes.bool                              [10, 53]                         None
+  55 Ops.CMPNE           :            dtypes.bool                              [53, 53]                         None
+  56 Ops.WHERE           :            dtypes.float                             [55, 53, 10]                     None
+  57 Ops.WHERE           :            dtypes.float                             [54, 53, 56]                     None
+  58 Ops.STORE           :            dtypes.void                              [7, 57]                          None
+  59 Ops.ADD             :            dtypes.int                               [6, 31]                          None
+  60 Ops.INDEX           :            dtypes.float                             [0, 59]                          None
+  61 Ops.CONST           :            dtypes.weakint                           []                               7
+  62 Ops.CAST            :            dtypes.int                               ['7']                            dtypes.int
+  63 Ops.INDEX           :            dtypes.bool                              [11, 62]                         None
+  64 Ops.LOAD            :            dtypes.bool                              [63]                             None
+  65 Ops.ADD             :            dtypes.int                               [17, 19]                         None
+  66 Ops.SHRINK          :            dtypes.float                             [14, 65, 19]                     None
+  67 Ops.LOAD            :            dtypes.float                             [66]                             None
+  68 Ops.INDEX           :            dtypes.float                             [67, 5]                          None
+  69 Ops.WHERE           :            dtypes.float                             [64, 68, 24]                     None
+  70 Ops.CONST           :            dtypes.weakint                           []                               6
+  71 Ops.CAST            :            dtypes.int                               ['6']                            dtypes.int
+  72 Ops.INDEX           :            dtypes.bool                              [11, 71]                         None
+  73 Ops.LOAD            :            dtypes.bool                              [72]                             None
+  74 Ops.INDEX           :            dtypes.float                             [67, 2]                          None
+  75 Ops.WHERE           :            dtypes.float                             [73, 74, 24]                     None
+  76 Ops.CONST           :            dtypes.weakint                           []                               5
+  77 Ops.CAST            :            dtypes.int                               ['5']                            dtypes.int
+  78 Ops.INDEX           :            dtypes.bool                              [11, 77]                         None
   79 Ops.LOAD            :            dtypes.bool                              [78]                             None
-  80 Ops.INDEX           :            dtypes.float                             [57, 5]                          None
-  81 Ops.WHERE           :            dtypes.float                             [79, 80, 23]                     None
-  82 Ops.CMPLT           :            dtypes.bool                              [75, 81]                         None
-  83 Ops.WHERE           :            dtypes.float                             [82, 81, 75]                     None
-  84 Ops.CMPLT           :            dtypes.bool                              [83, 47]                         None
-  85 Ops.WHERE           :            dtypes.float                             [84, 47, 83]                     None
-  86 Ops.STORE           :            dtypes.void                              [52, 85]                         None
-  87 Ops.ADD             :            dtypes.int                               [6, 2]                           None
-  88 Ops.INDEX           :            dtypes.float                             [0, 87]                          None
-  89 Ops.CONST           :            dtypes.weakint                           []                               8
-  90 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-  91 Ops.INDEX           :            dtypes.bool                              [8, 90]                          None
-  92 Ops.LOAD            :            dtypes.bool                              [91]                             None
-  93 Ops.ADD             :            dtypes.int                               [16, 90]                         None
-  94 Ops.SHRINK          :            dtypes.float                             [13, 93, 18]                     None
-  95 Ops.LOAD            :            dtypes.float                             [94]                             None
-  96 Ops.INDEX           :            dtypes.float                             [95, 10]                         None
-  97 Ops.WHERE           :            dtypes.float                             [92, 96, 23]                     None
-  98 Ops.CONST           :            dtypes.weakint                           []                               9
-  99 Ops.CAST            :            dtypes.int                               ['9']                            dtypes.int
- 100 Ops.INDEX           :            dtypes.bool                              [8, 99]                          None
- 101 Ops.LOAD            :            dtypes.bool                              [100]                            None
- 102 Ops.INDEX           :            dtypes.float                             [95, 26]                         None
- 103 Ops.WHERE           :            dtypes.float                             [101, 102, 23]                   None
- 104 Ops.CMPLT           :            dtypes.bool                              [97, 103]                        None
- 105 Ops.WHERE           :            dtypes.float                             [104, 103, 97]                   None
- 106 Ops.CONST           :            dtypes.weakint                           []                               10
- 107 Ops.CAST            :            dtypes.int                               ['10']                           dtypes.int
- 108 Ops.INDEX           :            dtypes.bool                              [8, 107]                         None
- 109 Ops.LOAD            :            dtypes.bool                              [108]                            None
- 110 Ops.INDEX           :            dtypes.float                             [95, 2]                          None
- 111 Ops.WHERE           :            dtypes.float                             [109, 110, 23]                   None
- 112 Ops.CMPLT           :            dtypes.bool                              [105, 111]                       None
- 113 Ops.WHERE           :            dtypes.float                             [112, 111, 105]                  None
- 114 Ops.CONST           :            dtypes.weakint                           []                               11
- 115 Ops.CAST            :            dtypes.int                               ['11']                           dtypes.int
- 116 Ops.INDEX           :            dtypes.bool                              [8, 115]                         None
- 117 Ops.LOAD            :            dtypes.bool                              [116]                            None
- 118 Ops.INDEX           :            dtypes.float                             [95, 5]                          None
- 119 Ops.WHERE           :            dtypes.float                             [117, 118, 23]                   None
- 120 Ops.CMPLT           :            dtypes.bool                              [113, 119]                       None
- 121 Ops.WHERE           :            dtypes.float                             [120, 119, 113]                  None
- 122 Ops.CMPLT           :            dtypes.bool                              [121, 47]                        None
- 123 Ops.WHERE           :            dtypes.float                             [122, 47, 121]                   None
- 124 Ops.STORE           :            dtypes.void                              [88, 123]                        None
- 125 Ops.GROUP           :            dtypes.void                              [50, 86, 124]                    None
- 126 Ops.SINK            :            dtypes.void                              [125]                            KernelInfo(name='r_2_3_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=2, arg=(0, AxisType.UNROLL)), Opt(op=OptOps.SPLIT, axis=0, arg=(2, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
+  80 Ops.INDEX           :            dtypes.float                             [67, 31]                         None
+  81 Ops.WHERE           :            dtypes.float                             [79, 80, 24]                     None
+  82 Ops.INDEX           :            dtypes.bool                              [11, 19]                         None
+  83 Ops.LOAD            :            dtypes.bool                              [82]                             None
+  84 Ops.INDEX           :            dtypes.float                             [67, 37]                         None
+  85 Ops.WHERE           :            dtypes.float                             [83, 84, 24]                     None
+  86 Ops.CMPLT           :            dtypes.bool                              [81, 85]                         None
+  87 Ops.CMPNE           :            dtypes.bool                              [85, 85]                         None
+  88 Ops.WHERE           :            dtypes.float                             [87, 85, 81]                     None
+  89 Ops.WHERE           :            dtypes.float                             [86, 85, 88]                     None
+  90 Ops.CMPLT           :            dtypes.bool                              [75, 89]                         None
+  91 Ops.CMPNE           :            dtypes.bool                              [89, 89]                         None
+  92 Ops.WHERE           :            dtypes.float                             [91, 89, 75]                     None
+  93 Ops.WHERE           :            dtypes.float                             [90, 89, 92]                     None
+  94 Ops.CMPLT           :            dtypes.bool                              [69, 93]                         None
+  95 Ops.CMPNE           :            dtypes.bool                              [93, 93]                         None
+  96 Ops.WHERE           :            dtypes.float                             [95, 93, 69]                     None
+  97 Ops.WHERE           :            dtypes.float                             [94, 93, 96]                     None
+  98 Ops.CMPLT           :            dtypes.bool                              [10, 97]                         None
+  99 Ops.CMPNE           :            dtypes.bool                              [97, 97]                         None
+ 100 Ops.WHERE           :            dtypes.float                             [99, 97, 10]                     None
+ 101 Ops.WHERE           :            dtypes.float                             [98, 97, 100]                    None
+ 102 Ops.STORE           :            dtypes.void                              [60, 101]                        None
+ 103 Ops.ADD             :            dtypes.int                               [6, 2]                           None
+ 104 Ops.INDEX           :            dtypes.float                             [0, 103]                         None
+ 105 Ops.CONST           :            dtypes.weakint                           []                               11
+ 106 Ops.CAST            :            dtypes.int                               ['11']                           dtypes.int
+ 107 Ops.INDEX           :            dtypes.bool                              [11, 106]                        None
+ 108 Ops.LOAD            :            dtypes.bool                              [107]                            None
+ 109 Ops.CONST           :            dtypes.weakint                           []                               8
+ 110 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+ 111 Ops.ADD             :            dtypes.int                               [17, 110]                        None
+ 112 Ops.SHRINK          :            dtypes.float                             [14, 111, 19]                    None
+ 113 Ops.LOAD            :            dtypes.float                             [112]                            None
+ 114 Ops.INDEX           :            dtypes.float                             [113, 5]                         None
+ 115 Ops.WHERE           :            dtypes.float                             [108, 114, 24]                   None
+ 116 Ops.CONST           :            dtypes.weakint                           []                               10
+ 117 Ops.CAST            :            dtypes.int                               ['10']                           dtypes.int
+ 118 Ops.INDEX           :            dtypes.bool                              [11, 117]                        None
+ 119 Ops.LOAD            :            dtypes.bool                              [118]                            None
+ 120 Ops.INDEX           :            dtypes.float                             [113, 2]                         None
+ 121 Ops.WHERE           :            dtypes.float                             [119, 120, 24]                   None
+ 122 Ops.CONST           :            dtypes.weakint                           []                               9
+ 123 Ops.CAST            :            dtypes.int                               ['9']                            dtypes.int
+ 124 Ops.INDEX           :            dtypes.bool                              [11, 123]                        None
+ 125 Ops.LOAD            :            dtypes.bool                              [124]                            None
+ 126 Ops.INDEX           :            dtypes.float                             [113, 31]                        None
+ 127 Ops.WHERE           :            dtypes.float                             [125, 126, 24]                   None
+ 128 Ops.INDEX           :            dtypes.bool                              [11, 110]                        None
+ 129 Ops.LOAD            :            dtypes.bool                              [128]                            None
+ 130 Ops.INDEX           :            dtypes.float                             [113, 37]                        None
+ 131 Ops.WHERE           :            dtypes.float                             [129, 130, 24]                   None
+ 132 Ops.CMPLT           :            dtypes.bool                              [127, 131]                       None
+ 133 Ops.CMPNE           :            dtypes.bool                              [131, 131]                       None
+ 134 Ops.WHERE           :            dtypes.float                             [133, 131, 127]                  None
+ 135 Ops.WHERE           :            dtypes.float                             [132, 131, 134]                  None
+ 136 Ops.CMPLT           :            dtypes.bool                              [121, 135]                       None
+ 137 Ops.CMPNE           :            dtypes.bool                              [135, 135]                       None
+ 138 Ops.WHERE           :            dtypes.float                             [137, 135, 121]                  None
+ 139 Ops.WHERE           :            dtypes.float                             [136, 135, 138]                  None
+ 140 Ops.CMPLT           :            dtypes.bool                              [115, 139]                       None
+ 141 Ops.CMPNE           :            dtypes.bool                              [139, 139]                       None
+ 142 Ops.WHERE           :            dtypes.float                             [141, 139, 115]                  None
+ 143 Ops.WHERE           :            dtypes.float                             [140, 139, 142]                  None
+ 144 Ops.CMPLT           :            dtypes.bool                              [10, 143]                        None
+ 145 Ops.CMPNE           :            dtypes.bool                              [143, 143]                       None
+ 146 Ops.WHERE           :            dtypes.float                             [145, 143, 10]                   None
+ 147 Ops.WHERE           :            dtypes.float                             [144, 143, 146]                  None
+ 148 Ops.STORE           :            dtypes.void                              [104, 147]                       None
+ 149 Ops.GROUP           :            dtypes.void                              [58, 102, 148]                   None
+ 150 Ops.SINK            :            dtypes.void                              [149]                            KernelInfo(name='r_2_3_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=2, arg=(0, AxisType.UNROLL)), Opt(op=OptOps.SPLIT, axis=0, arg=(2, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 1 ===
    0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 6, device='CPU')
    1 Ops.CONST           :            dtypes.weakint                           []                               2
@@ -135,167 +159,172 @@
    5 Ops.CAST            :            dtypes.int                               ['3']                            dtypes.int
    6 Ops.MUL             :            dtypes.int                               [3, 5]                           None
    7 Ops.INDEX           :            dtypes.float                             [0, 6]                           None
-   8 Ops.PARAM           :            dtypes.bool                              []                               ParamArg(1, dtypes.bool, 12, device='CPU')
-   9 Ops.CONST           :            dtypes.weakint                           []                               0
-  10 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
-  11 Ops.INDEX           :            dtypes.bool                              [8, 10]                          None
-  12 Ops.LOAD            :            dtypes.bool                              [11]                             None
-  13 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 24, device='CPU')
-  14 Ops.CONST           :            dtypes.weakint                           []                               12
-  15 Ops.CAST            :            dtypes.int                               ['12']                           dtypes.int
-  16 Ops.MUL             :            dtypes.int                               [3, 15]                          None
-  17 Ops.CONST           :            dtypes.weakint                           []                               4
-  18 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-  19 Ops.SHRINK          :            dtypes.float                             [13, 16, 18]                     None
-  20 Ops.LOAD            :            dtypes.float                             [19]                             None
-  21 Ops.INDEX           :            dtypes.float                             [20, 10]                         None
-  22 Ops.CONST           :            dtypes.weakfloat                         []                               -inf
-  23 Ops.CAST            :            dtypes.float                             ['-inf']                         dtypes.float
-  24 Ops.WHERE           :            dtypes.float                             [12, 21, 23]                     None
-  25 Ops.PARAM           :            dtypes.float                             []                               ParamArg(3, dtypes.float, 6, device='CPU')
-  26 Ops.INDEX           :            dtypes.float                             [25, 6]                          None
-  27 Ops.LOAD            :            dtypes.float                             [26]                             None
-  28 Ops.SUB             :            dtypes.float                             [24, 27]                         None
-  29 Ops.CONST           :            dtypes.weakfloat                         []                               1.4426950216293335
-  30 Ops.CAST            :            dtypes.float                             ['1.4426950216293335']           dtypes.float
-  31 Ops.MUL             :            dtypes.float                             [28, 30]                         None
-  32 Ops.EXP2            :            dtypes.float                             [31]                             None
-  33 Ops.CONST           :            dtypes.weakint                           []                               1
-  34 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
-  35 Ops.INDEX           :            dtypes.bool                              [8, 34]                          None
-  36 Ops.LOAD            :            dtypes.bool                              [35]                             None
-  37 Ops.INDEX           :            dtypes.float                             [20, 34]                         None
-  38 Ops.WHERE           :            dtypes.float                             [36, 37, 23]                     None
-  39 Ops.SUB             :            dtypes.float                             [38, 27]                         None
-  40 Ops.MUL             :            dtypes.float                             [39, 30]                         None
-  41 Ops.EXP2            :            dtypes.float                             [40]                             None
-  42 Ops.ADD             :            dtypes.float                             [32, 41]                         None
-  43 Ops.INDEX           :            dtypes.bool                              [8, 2]                           None
-  44 Ops.LOAD            :            dtypes.bool                              [43]                             None
-  45 Ops.INDEX           :            dtypes.float                             [20, 2]                          None
-  46 Ops.WHERE           :            dtypes.float                             [44, 45, 23]                     None
-  47 Ops.SUB             :            dtypes.float                             [46, 27]                         None
-  48 Ops.MUL             :            dtypes.float                             [47, 30]                         None
-  49 Ops.EXP2            :            dtypes.float                             [48]                             None
-  50 Ops.ADD             :            dtypes.float                             [42, 49]                         None
-  51 Ops.INDEX           :            dtypes.bool                              [8, 5]                           None
-  52 Ops.LOAD            :            dtypes.bool                              [51]                             None
-  53 Ops.INDEX           :            dtypes.float                             [20, 5]                          None
-  54 Ops.WHERE           :            dtypes.float                             [52, 53, 23]                     None
-  55 Ops.SUB             :            dtypes.float                             [54, 27]                         None
-  56 Ops.MUL             :            dtypes.float                             [55, 30]                         None
-  57 Ops.EXP2            :            dtypes.float                             [56]                             None
-  58 Ops.ADD             :            dtypes.float                             [50, 57]                         None
-  59 Ops.PARAM           :            dtypes.float                             []                               ParamArg(4, dtypes.float, 2, device='CPU')
-  60 Ops.INDEX           :            dtypes.float                             [59, 3]                          None
-  61 Ops.LOAD            :            dtypes.float                             [60]                             None
-  62 Ops.SUB             :            dtypes.float                             [61, 27]                         None
-  63 Ops.MUL             :            dtypes.float                             [62, 30]                         None
-  64 Ops.EXP2            :            dtypes.float                             [63]                             None
-  65 Ops.ADD             :            dtypes.float                             [58, 64]                         None
-  66 Ops.STORE           :            dtypes.void                              [7, 65]                          None
-  67 Ops.ADD             :            dtypes.int                               [6, 34]                          None
-  68 Ops.INDEX           :            dtypes.float                             [0, 67]                          None
-  69 Ops.INDEX           :            dtypes.bool                              [8, 18]                          None
-  70 Ops.LOAD            :            dtypes.bool                              [69]                             None
-  71 Ops.ADD             :            dtypes.int                               [16, 18]                         None
-  72 Ops.SHRINK          :            dtypes.float                             [13, 71, 18]                     None
-  73 Ops.LOAD            :            dtypes.float                             [72]                             None
-  74 Ops.INDEX           :            dtypes.float                             [73, 10]                         None
-  75 Ops.WHERE           :            dtypes.float                             [70, 74, 23]                     None
-  76 Ops.INDEX           :            dtypes.float                             [25, 67]                         None
-  77 Ops.LOAD            :            dtypes.float                             [76]                             None
-  78 Ops.SUB             :            dtypes.float                             [75, 77]                         None
-  79 Ops.MUL             :            dtypes.float                             [78, 30]                         None
-  80 Ops.EXP2            :            dtypes.float                             [79]                             None
-  81 Ops.CONST           :            dtypes.weakint                           []                               5
-  82 Ops.CAST            :            dtypes.int                               ['5']                            dtypes.int
-  83 Ops.INDEX           :            dtypes.bool                              [8, 82]                          None
-  84 Ops.LOAD            :            dtypes.bool                              [83]                             None
-  85 Ops.INDEX           :            dtypes.float                             [73, 34]                         None
-  86 Ops.WHERE           :            dtypes.float                             [84, 85, 23]                     None
-  87 Ops.SUB             :            dtypes.float                             [86, 77]                         None
-  88 Ops.MUL             :            dtypes.float                             [87, 30]                         None
-  89 Ops.EXP2            :            dtypes.float                             [88]                             None
-  90 Ops.ADD             :            dtypes.float                             [80, 89]                         None
-  91 Ops.CONST           :            dtypes.weakint                           []                               6
-  92 Ops.CAST            :            dtypes.int                               ['6']                            dtypes.int
-  93 Ops.INDEX           :            dtypes.bool                              [8, 92]                          None
-  94 Ops.LOAD            :            dtypes.bool                              [93]                             None
-  95 Ops.INDEX           :            dtypes.float                             [73, 2]                          None
-  96 Ops.WHERE           :            dtypes.float                             [94, 95, 23]                     None
-  97 Ops.SUB             :            dtypes.float                             [96, 77]                         None
-  98 Ops.MUL             :            dtypes.float                             [97, 30]                         None
-  99 Ops.EXP2            :            dtypes.float                             [98]                             None
- 100 Ops.ADD             :            dtypes.float                             [90, 99]                         None
- 101 Ops.CONST           :            dtypes.weakint                           []                               7
- 102 Ops.CAST            :            dtypes.int                               ['7']                            dtypes.int
- 103 Ops.INDEX           :            dtypes.bool                              [8, 102]                         None
- 104 Ops.LOAD            :            dtypes.bool                              [103]                            None
- 105 Ops.INDEX           :            dtypes.float                             [73, 5]                          None
- 106 Ops.WHERE           :            dtypes.float                             [104, 105, 23]                   None
- 107 Ops.SUB             :            dtypes.float                             [106, 77]                        None
- 108 Ops.MUL             :            dtypes.float                             [107, 30]                        None
- 109 Ops.EXP2            :            dtypes.float                             [108]                            None
- 110 Ops.ADD             :            dtypes.float                             [100, 109]                       None
- 111 Ops.SUB             :            dtypes.float                             [61, 77]                         None
- 112 Ops.MUL             :            dtypes.float                             [111, 30]                        None
+   8 Ops.CONST           :            dtypes.weakfloat                         []                               0.0
+   9 Ops.CAST            :            dtypes.float                             ['0.0']                          dtypes.float
+  10 Ops.PARAM           :            dtypes.bool                              []                               ParamArg(1, dtypes.bool, 12, device='CPU')
+  11 Ops.CONST           :            dtypes.weakint                           []                               0
+  12 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
+  13 Ops.INDEX           :            dtypes.bool                              [10, 12]                         None
+  14 Ops.LOAD            :            dtypes.bool                              [13]                             None
+  15 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 24, device='CPU')
+  16 Ops.CONST           :            dtypes.weakint                           []                               12
+  17 Ops.CAST            :            dtypes.int                               ['12']                           dtypes.int
+  18 Ops.MUL             :            dtypes.int                               [3, 17]                          None
+  19 Ops.CONST           :            dtypes.weakint                           []                               4
+  20 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+  21 Ops.SHRINK          :            dtypes.float                             [15, 18, 20]                     None
+  22 Ops.LOAD            :            dtypes.float                             [21]                             None
+  23 Ops.INDEX           :            dtypes.float                             [22, 12]                         None
+  24 Ops.CONST           :            dtypes.weakfloat                         []                               -inf
+  25 Ops.CAST            :            dtypes.float                             ['-inf']                         dtypes.float
+  26 Ops.WHERE           :            dtypes.float                             [14, 23, 25]                     None
+  27 Ops.PARAM           :            dtypes.float                             []                               ParamArg(3, dtypes.float, 6, device='CPU')
+  28 Ops.INDEX           :            dtypes.float                             [27, 6]                          None
+  29 Ops.LOAD            :            dtypes.float                             [28]                             None
+  30 Ops.SUB             :            dtypes.float                             [26, 29]                         None
+  31 Ops.CONST           :            dtypes.weakfloat                         []                               1.4426950216293335
+  32 Ops.CAST            :            dtypes.float                             ['1.4426950216293335']           dtypes.float
+  33 Ops.MUL             :            dtypes.float                             [30, 32]                         None
+  34 Ops.EXP2            :            dtypes.float                             [33]                             None
+  35 Ops.ADD             :            dtypes.float                             [9, 34]                          None
+  36 Ops.CONST           :            dtypes.weakint                           []                               1
+  37 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
+  38 Ops.INDEX           :            dtypes.bool                              [10, 37]                         None
+  39 Ops.LOAD            :            dtypes.bool                              [38]                             None
+  40 Ops.INDEX           :            dtypes.float                             [22, 37]                         None
+  41 Ops.WHERE           :            dtypes.float                             [39, 40, 25]                     None
+  42 Ops.SUB             :            dtypes.float                             [41, 29]                         None
+  43 Ops.MUL             :            dtypes.float                             [42, 32]                         None
+  44 Ops.EXP2            :            dtypes.float                             [43]                             None
+  45 Ops.ADD             :            dtypes.float                             [35, 44]                         None
+  46 Ops.INDEX           :            dtypes.bool                              [10, 2]                          None
+  47 Ops.LOAD            :            dtypes.bool                              [46]                             None
+  48 Ops.INDEX           :            dtypes.float                             [22, 2]                          None
+  49 Ops.WHERE           :            dtypes.float                             [47, 48, 25]                     None
+  50 Ops.SUB             :            dtypes.float                             [49, 29]                         None
+  51 Ops.MUL             :            dtypes.float                             [50, 32]                         None
+  52 Ops.EXP2            :            dtypes.float                             [51]                             None
+  53 Ops.ADD             :            dtypes.float                             [45, 52]                         None
+  54 Ops.INDEX           :            dtypes.bool                              [10, 5]                          None
+  55 Ops.LOAD            :            dtypes.bool                              [54]                             None
+  56 Ops.INDEX           :            dtypes.float                             [22, 5]                          None
+  57 Ops.WHERE           :            dtypes.float                             [55, 56, 25]                     None
+  58 Ops.SUB             :            dtypes.float                             [57, 29]                         None
+  59 Ops.MUL             :            dtypes.float                             [58, 32]                         None
+  60 Ops.EXP2            :            dtypes.float                             [59]                             None
+  61 Ops.ADD             :            dtypes.float                             [53, 60]                         None
+  62 Ops.PARAM           :            dtypes.float                             []                               ParamArg(4, dtypes.float, 2, device='CPU')
+  63 Ops.INDEX           :            dtypes.float                             [62, 3]                          None
+  64 Ops.LOAD            :            dtypes.float                             [63]                             None
+  65 Ops.SUB             :            dtypes.float                             [64, 29]                         None
+  66 Ops.MUL             :            dtypes.float                             [65, 32]                         None
+  67 Ops.EXP2            :            dtypes.float                             [66]                             None
+  68 Ops.ADD             :            dtypes.float                             [61, 67]                         None
+  69 Ops.STORE           :            dtypes.void                              [7, 68]                          None
+  70 Ops.ADD             :            dtypes.int                               [6, 37]                          None
+  71 Ops.INDEX           :            dtypes.float                             [0, 70]                          None
+  72 Ops.INDEX           :            dtypes.bool                              [10, 20]                         None
+  73 Ops.LOAD            :            dtypes.bool                              [72]                             None
+  74 Ops.ADD             :            dtypes.int                               [18, 20]                         None
+  75 Ops.SHRINK          :            dtypes.float                             [15, 74, 20]                     None
+  76 Ops.LOAD            :            dtypes.float                             [75]                             None
+  77 Ops.INDEX           :            dtypes.float                             [76, 12]                         None
+  78 Ops.WHERE           :            dtypes.float                             [73, 77, 25]                     None
+  79 Ops.INDEX           :            dtypes.float                             [27, 70]                         None
+  80 Ops.LOAD            :            dtypes.float                             [79]                             None
+  81 Ops.SUB             :            dtypes.float                             [78, 80]                         None
+  82 Ops.MUL             :            dtypes.float                             [81, 32]                         None
+  83 Ops.EXP2            :            dtypes.float                             [82]                             None
+  84 Ops.ADD             :            dtypes.float                             [9, 83]                          None
+  85 Ops.CONST           :            dtypes.weakint                           []                               5
+  86 Ops.CAST            :            dtypes.int                               ['5']                            dtypes.int
+  87 Ops.INDEX           :            dtypes.bool                              [10, 86]                         None
+  88 Ops.LOAD            :            dtypes.bool                              [87]                             None
+  89 Ops.INDEX           :            dtypes.float                             [76, 37]                         None
+  90 Ops.WHERE           :            dtypes.float                             [88, 89, 25]                     None
+  91 Ops.SUB             :            dtypes.float                             [90, 80]                         None
+  92 Ops.MUL             :            dtypes.float                             [91, 32]                         None
+  93 Ops.EXP2            :            dtypes.float                             [92]                             None
+  94 Ops.ADD             :            dtypes.float                             [84, 93]                         None
+  95 Ops.CONST           :            dtypes.weakint                           []                               6
+  96 Ops.CAST            :            dtypes.int                               ['6']                            dtypes.int
+  97 Ops.INDEX           :            dtypes.bool                              [10, 96]                         None
+  98 Ops.LOAD            :            dtypes.bool                              [97]                             None
+  99 Ops.INDEX           :            dtypes.float                             [76, 2]                          None
+ 100 Ops.WHERE           :            dtypes.float                             [98, 99, 25]                     None
+ 101 Ops.SUB             :            dtypes.float                             [100, 80]                        None
+ 102 Ops.MUL             :            dtypes.float                             [101, 32]                        None
+ 103 Ops.EXP2            :            dtypes.float                             [102]                            None
+ 104 Ops.ADD             :            dtypes.float                             [94, 103]                        None
+ 105 Ops.CONST           :            dtypes.weakint                           []                               7
+ 106 Ops.CAST            :            dtypes.int                               ['7']                            dtypes.int
+ 107 Ops.INDEX           :            dtypes.bool                              [10, 106]                        None
+ 108 Ops.LOAD            :            dtypes.bool                              [107]                            None
+ 109 Ops.INDEX           :            dtypes.float                             [76, 5]                          None
+ 110 Ops.WHERE           :            dtypes.float                             [108, 109, 25]                   None
+ 111 Ops.SUB             :            dtypes.float                             [110, 80]                        None
+ 112 Ops.MUL             :            dtypes.float                             [111, 32]                        None
  113 Ops.EXP2            :            dtypes.float                             [112]                            None
- 114 Ops.ADD             :            dtypes.float                             [110, 113]                       None
- 115 Ops.STORE           :            dtypes.void                              [68, 114]                        None
- 116 Ops.ADD             :            dtypes.int                               [6, 2]                           None
- 117 Ops.INDEX           :            dtypes.float                             [0, 116]                         None
- 118 Ops.CONST           :            dtypes.weakint                           []                               8
- 119 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
- 120 Ops.INDEX           :            dtypes.bool                              [8, 119]                         None
- 121 Ops.LOAD            :            dtypes.bool                              [120]                            None
- 122 Ops.ADD             :            dtypes.int                               [16, 119]                        None
- 123 Ops.SHRINK          :            dtypes.float                             [13, 122, 18]                    None
- 124 Ops.LOAD            :            dtypes.float                             [123]                            None
- 125 Ops.INDEX           :            dtypes.float                             [124, 10]                        None
- 126 Ops.WHERE           :            dtypes.float                             [121, 125, 23]                   None
- 127 Ops.INDEX           :            dtypes.float                             [25, 116]                        None
+ 114 Ops.ADD             :            dtypes.float                             [104, 113]                       None
+ 115 Ops.SUB             :            dtypes.float                             [64, 80]                         None
+ 116 Ops.MUL             :            dtypes.float                             [115, 32]                        None
+ 117 Ops.EXP2            :            dtypes.float                             [116]                            None
+ 118 Ops.ADD             :            dtypes.float                             [114, 117]                       None
+ 119 Ops.STORE           :            dtypes.void                              [71, 118]                        None
+ 120 Ops.ADD             :            dtypes.int                               [6, 2]                           None
+ 121 Ops.INDEX           :            dtypes.float                             [0, 120]                         None
+ 122 Ops.CONST           :            dtypes.weakint                           []                               8
+ 123 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+ 124 Ops.INDEX           :            dtypes.bool                              [10, 123]                        None
+ 125 Ops.LOAD            :            dtypes.bool                              [124]                            None
+ 126 Ops.ADD             :            dtypes.int                               [18, 123]                        None
+ 127 Ops.SHRINK          :            dtypes.float                             [15, 126, 20]                    None
  128 Ops.LOAD            :            dtypes.float                             [127]                            None
- 129 Ops.SUB             :            dtypes.float                             [126, 128]                       None
- 130 Ops.MUL             :            dtypes.float                             [129, 30]                        None
- 131 Ops.EXP2            :            dtypes.float                             [130]                            None
- 132 Ops.CONST           :            dtypes.weakint                           []                               9
- 133 Ops.CAST            :            dtypes.int                               ['9']                            dtypes.int
- 134 Ops.INDEX           :            dtypes.bool                              [8, 133]                         None
- 135 Ops.LOAD            :            dtypes.bool                              [134]                            None
- 136 Ops.INDEX           :            dtypes.float                             [124, 34]                        None
- 137 Ops.WHERE           :            dtypes.float                             [135, 136, 23]                   None
- 138 Ops.SUB             :            dtypes.float                             [137, 128]                       None
- 139 Ops.MUL             :            dtypes.float                             [138, 30]                        None
- 140 Ops.EXP2            :            dtypes.float                             [139]                            None
- 141 Ops.ADD             :            dtypes.float                             [131, 140]                       None
- 142 Ops.CONST           :            dtypes.weakint                           []                               10
- 143 Ops.CAST            :            dtypes.int                               ['10']                           dtypes.int
- 144 Ops.INDEX           :            dtypes.bool                              [8, 143]                         None
- 145 Ops.LOAD            :            dtypes.bool                              [144]                            None
- 146 Ops.INDEX           :            dtypes.float                             [124, 2]                         None
- 147 Ops.WHERE           :            dtypes.float                             [145, 146, 23]                   None
- 148 Ops.SUB             :            dtypes.float                             [147, 128]                       None
- 149 Ops.MUL             :            dtypes.float                             [148, 30]                        None
- 150 Ops.EXP2            :            dtypes.float                             [149]                            None
- 151 Ops.ADD             :            dtypes.float                             [141, 150]                       None
- 152 Ops.CONST           :            dtypes.weakint                           []                               11
- 153 Ops.CAST            :            dtypes.int                               ['11']                           dtypes.int
- 154 Ops.INDEX           :            dtypes.bool                              [8, 153]                         None
- 155 Ops.LOAD            :            dtypes.bool                              [154]                            None
- 156 Ops.INDEX           :            dtypes.float                             [124, 5]                         None
- 157 Ops.WHERE           :            dtypes.float                             [155, 156, 23]                   None
- 158 Ops.SUB             :            dtypes.float                             [157, 128]                       None
- 159 Ops.MUL             :            dtypes.float                             [158, 30]                        None
- 160 Ops.EXP2            :            dtypes.float                             [159]                            None
- 161 Ops.ADD             :            dtypes.float                             [151, 160]                       None
- 162 Ops.SUB             :            dtypes.float                             [61, 128]                        None
- 163 Ops.MUL             :            dtypes.float                             [162, 30]                        None
- 164 Ops.EXP2            :            dtypes.float                             [163]                            None
- 165 Ops.ADD             :            dtypes.float                             [161, 164]                       None
- 166 Ops.STORE           :            dtypes.void                              [117, 165]                       None
- 167 Ops.GROUP           :            dtypes.void                              [66, 115, 166]                   None
- 168 Ops.SINK            :            dtypes.void                              [167]                            KernelInfo(name='r_2_3_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=2, arg=(0, AxisType.UNROLL)), Opt(op=OptOps.SPLIT, axis=0, arg=(2, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
+ 129 Ops.INDEX           :            dtypes.float                             [128, 12]                        None
+ 130 Ops.WHERE           :            dtypes.float                             [125, 129, 25]                   None
+ 131 Ops.INDEX           :            dtypes.float                             [27, 120]                        None
+ 132 Ops.LOAD            :            dtypes.float                             [131]                            None
+ 133 Ops.SUB             :            dtypes.float                             [130, 132]                       None
+ 134 Ops.MUL             :            dtypes.float                             [133, 32]                        None
+ 135 Ops.EXP2            :            dtypes.float                             [134]                            None
+ 136 Ops.ADD             :            dtypes.float                             [9, 135]                         None
+ 137 Ops.CONST           :            dtypes.weakint                           []                               9
+ 138 Ops.CAST            :            dtypes.int                               ['9']                            dtypes.int
+ 139 Ops.INDEX           :            dtypes.bool                              [10, 138]                        None
+ 140 Ops.LOAD            :            dtypes.bool                              [139]                            None
+ 141 Ops.INDEX           :            dtypes.float                             [128, 37]                        None
+ 142 Ops.WHERE           :            dtypes.float                             [140, 141, 25]                   None
+ 143 Ops.SUB             :            dtypes.float                             [142, 132]                       None
+ 144 Ops.MUL             :            dtypes.float                             [143, 32]                        None
+ 145 Ops.EXP2            :            dtypes.float                             [144]                            None
+ 146 Ops.ADD             :            dtypes.float                             [136, 145]                       None
+ 147 Ops.CONST           :            dtypes.weakint                           []                               10
+ 148 Ops.CAST            :            dtypes.int                               ['10']                           dtypes.int
+ 149 Ops.INDEX           :            dtypes.bool                              [10, 148]                        None
+ 150 Ops.LOAD            :            dtypes.bool                              [149]                            None
+ 151 Ops.INDEX           :            dtypes.float                             [128, 2]                         None
+ 152 Ops.WHERE           :            dtypes.float                             [150, 151, 25]                   None
+ 153 Ops.SUB             :            dtypes.float                             [152, 132]                       None
+ 154 Ops.MUL             :            dtypes.float                             [153, 32]                        None
+ 155 Ops.EXP2            :            dtypes.float                             [154]                            None
+ 156 Ops.ADD             :            dtypes.float                             [146, 155]                       None
+ 157 Ops.CONST           :            dtypes.weakint                           []                               11
+ 158 Ops.CAST            :            dtypes.int                               ['11']                           dtypes.int
+ 159 Ops.INDEX           :            dtypes.bool                              [10, 158]                        None
+ 160 Ops.LOAD            :            dtypes.bool                              [159]                            None
+ 161 Ops.INDEX           :            dtypes.float                             [128, 5]                         None
+ 162 Ops.WHERE           :            dtypes.float                             [160, 161, 25]                   None
+ 163 Ops.SUB             :            dtypes.float                             [162, 132]                       None
+ 164 Ops.MUL             :            dtypes.float                             [163, 32]                        None
+ 165 Ops.EXP2            :            dtypes.float                             [164]                            None
+ 166 Ops.ADD             :            dtypes.float                             [156, 165]                       None
+ 167 Ops.SUB             :            dtypes.float                             [64, 132]                        None
+ 168 Ops.MUL             :            dtypes.float                             [167, 32]                        None
+ 169 Ops.EXP2            :            dtypes.float                             [168]                            None
+ 170 Ops.ADD             :            dtypes.float                             [166, 169]                       None
+ 171 Ops.STORE           :            dtypes.void                              [121, 170]                       None
+ 172 Ops.GROUP           :            dtypes.void                              [69, 119, 171]                   None
+ 173 Ops.SINK            :            dtypes.void                              [172]                            KernelInfo(name='r_2_3_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=2, arg=(0, AxisType.UNROLL)), Opt(op=OptOps.SPLIT, axis=0, arg=(2, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 2 ===
    0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 24, device='CPU')
    1 Ops.CONST           :            dtypes.weakint                           []                               2
